gateway-sp-comms/src/recv_handler/mod.rs

```diff
@@ -0,0 +1,200 @@
+    pub(crate) fn new(
+        switch_discovery: ManagementSwitchDiscovery,
+        log: Logger,
+    ) -> (ManagementSwitch, Arc<Self>) {
+        // prime `sp_state` with all known ports of the switch
+        let all_ports = switch_discovery.all_ports();
+        let mut sp_state = HashMap::with_capacity(all_ports.len());
+        for port in all_ports {
+            sp_state.insert(port, SingleSpState::default());
+        }
+
+        // configure a `ManagementSwitch` that notifies us of every incoming
+        // packet
+        let handler = Arc::new(Self { sp_state, log });
+        let switch = {
+            let handler = Arc::clone(&handler);
+            switch_discovery.start_recv_task(move |port, buf| {
+                handler.handle_incoming_packet(port, buf)
+            })
+        };
+
+        (switch, handler)
+    }
+
+    // SwitchPort instances can only be created by `ManagementSwitch`, so we
+    // should never be able to instantiate a port that we don't have in
+    // `self.sp_state` (which we initialize with all ports declared by the
+    // switch we were given).
+    fn sp_state(&self, port: SwitchPort) -> &SingleSpState {
+        self.sp_state.get(&port).expect("invalid switch port")
+    }
+
+    /// Get our current serial console contents for the given SP component.
+    pub(crate) fn serial_console_contents(
+        &self,
+        port: SwitchPort,
+        component: &SpComponent,
+    ) -> Option<SerialConsoleContents> {
+        self.sp_state(port).serial_console.lock().unwrap().contents(component)
+    }
+
+    /// Returns a future that will complete when we receive a response on the
+    /// given `port` with the corresponding `request_id`.
+    ///
+    /// Panics if `port` is not one of the ports defined by the `switch` given
+    /// to this `RecvHandler` when it was constructed.
+    pub(crate) async fn register_request_id(
+        &self,
+        port: SwitchPort,
+        request_id: u32,
+    ) -> Result<ResponseKind, ResponseError> {
+        self.sp_state(port).requests.wait_for_response(request_id).await
+    }
+
+    fn handle_incoming_packet(&self, port: SwitchPort, buf: &[u8]) {
+        trace!(&self.log, "received {} bytes from {:?}", buf.len(), port);
+
+        // parse into an `SpMessage`
+        let sp_msg = match gateway_messages::deserialize::<SpMessage>(buf) {
+            Ok((msg, _extra)) => {
+                // TODO should we check that `extra` is empty? if the
+                // response is maximal size any extra data is silently
+                // discarded anyway, so probably not?
+                msg
+            }
+            Err(err) => {
+                error!(&self.log, "discarding malformed message ({})", err);
+                return;
+            }
+        };
+        debug!(&self.log, "received {:?} from {:?}", sp_msg, port);
+
+        // `version` is intentionally the first 4 bytes of the packet; we
+        // could check it before trying to deserialize?
```

Comment:
yeah; I think we might kick ourselves if we don't..

Reply:
Changed in 93b27cd